clients/drcachesim/simulator/caching_device.cpp

```diff
@@ -152,19 +157,54 @@ caching_device_t::request(const memref_t &memref_in)
                 parent->stats->child_access(memref, false, cache_block);
                 parent->request(memref);
             }
+            if (snoop_filter != NULL) {
+                // Update snoop filter, other private caches invalidated on write.
+                snoop_filter->snoop(tag, id, (memref.data.type == TRACE_TYPE_WRITE));
+            }
 
-            // FIXME i#1726: coherence policy
-
+            addr_t victim_tag = cache_block->tag;
             // Check if we are inserting a new block, if we are then increment
             // the block loaded count.
-            if (cache_block->tag == TAG_INVALID) {
+            if (victim_tag == TAG_INVALID) {
                 loaded_blocks++;
-            } else if (inclusive && !children.empty()) {
-                for (auto &child : children) {
-                    child->invalidate(cache_block->tag);
+            } else {
+                if (!children.empty() && inclusive) {
+                    for (auto &child : children) {
+                        child->invalidate(victim_tag, INVALIDATION_INCLUSIVE);
+                    }
+                }
+                if (coherent_cache) {
+                    bool child_holds_tag = false;
+                    if (!children.empty()) {
+                        // We must check child caches to
+                        // find out if snoop filter may clear our bit.
+                        for (auto &child : children) {
+                            if (child->check_tags(victim_tag)) {
+                                child_holds_tag = true;
+                                break;
+                            }
+                        }
+                    }
+                    if (!child_holds_tag) {
+                        if (snoop_filter != NULL) {
+                            // Inform snoop filter of evicted line.
+                            snoop_filter->eviction_notification(victim_tag, id);
+                        } else if (parent != NULL) {
+                            // Inform parent of evicted line.
+                            parent->child_eviction(victim_tag, this);
+                        }
+                    }
                 }
             }
             cache_block->tag = tag;
+        } else if (coherent_cache && memref.data.type == TRACE_TYPE_WRITE) {
```

Comment:
nit: invert the parent `if` and bring this block up. Easier for readers to reason about the common case (hit) first.

Reply:
Fixed. Some hit stats logic is embedded in the tag search. Would it be beneficial to move this hit logic into this new if statement?

Reply:
Yes, that would be great. Then we can move the tag search into a small inline helper function, which just returns the block (or nullptr on a miss). As is, we have it replicated in 2-3 places.

Reply:
I consolidated the hit logic. Maybe the helper function should be a  separate fix?

Reply:
Sounds good.